### PR TITLE
Anikait - Get/Post CRUD Operations for Ride Backend

### LINKED ISSUE #45: Add GET/POST Operations for Ride Requests
Add Get/Post Operations with the following paths for Ride Requests
`/ride_request?id={} `- get a ride with given id (returns ride object iff user_id matches for when user not admin/driver)
`/ride_request/all`- get all rides (of given user id if user is not admin/driver, else all)
`/ride_request/post` Create a ride request with the given parameters of a Ride object except riderId

### PR DESCRIPTION
**Pull request by strongani**
---- EDIT ----
new version of this PR: #49

Added Get/Post CRUD Operations for Ride Backend, with the following api:
`/ride_request?id={} `- get a ride with given id, returns ride object iff user_id matches
`/ride_request/admin?id={} `- get a ride with given id, returns ride object no matter what (admin only)
`/ride_request/all`- get all rides of given user id
`/ride_request/admin/all`- get all rides across all users
`/ride_request/post` Create a ride request with the given parameters of a Ride object except riderId

Changed the Repository, Controller, and Controller Tests for the Ride Entity, following proj-courses PersonalSchedule entity as a guideline. Also changes ShiftRepository as a HotFix for maven backend.

Closes #45 and contributes to #7 

Swagger works as expected on LocalHost: (Note in this case, user is an admin and has Ride objects 2, 3, 4, so the `ride_request/all` only returns his rides, while `ride_request/admin/all` returns his rides along with ride object 5)

https://github.com/ucsb-cs156-s23/proj-gauchoride-s23-5pm-4/assets/29584856/687b60c7-89d5-45f3-8191-640718dbe5c1

### CODE REVIEW

File: src/test/java/edu/ucsb/cs156/gauchoride/controllers/RideControllerTests.java [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+import org.springframework.context.annotation.Import;
+import org.springframework.http.MediaType;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.web.servlet.MvcResult;
+
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
+
+// import java.time.LocalDateTime;
```
Comment (SophiaWeiler) [inline src/test/java/edu/ucsb/cs156/gauchoride/controllers/RideControllerTests.java]: Don't need this line... should probably remove it

File: src/main/java/edu/ucsb/cs156/gauchoride/controllers/RideController.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+@RequestMapping("/api/ride_request")
+@RestController
+
+public class RideController extends ApiController {
+
+    @Autowired
+    RideRepository rideRepository;
+
+    @ApiOperation(value = "List all rides")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (SophiaWeiler) [inline src/main/java/edu/ucsb/cs156/gauchoride/controllers/RideController.java]: should this be ROLE_DRIVER? drivers need to be able to see all of the rides as well

File: src/test/java/edu/ucsb/cs156/gauchoride/controllers/RideControllerTests.java [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+        @Test
+        public void logged_in_users_cannot_get_by_id_in_simple_test() throws Exception {
+                mockMvc.perform(get("/api/ride_request?id=7"))
+                                .andExpect(status().is(404)); 
+                // logged in users can get their rides by id unless their ride matches, no ride with id 7 exists
+        }
+
+
+        // Authorization tests for /api/ride_request/post
+        // (Perhaps should also have these for put and delete)
```
Comment (SophiaWeiler) [inline src/test/java/edu/ucsb/cs156/gauchoride/controllers/RideControllerTests.java]: I think we should get rid of this too, just make another issue

**Review by SophiaWeiler — CHANGES_REQUESTED** [review]
For the most part it looks good, there are just a few comments I made on the code that should be addressed before merging

File: src/main/java/edu/ucsb/cs156/gauchoride/controllers/RideController.java (hunk 2 above)
Comment (strongani) [reply src/main/java/edu/ucsb/cs156/gauchoride/controllers/RideController.java]: Edited!

**Review by SophiaWeiler — DISMISSED** [review]
lgtm now!

File: src/main/java/edu/ucsb/cs156/gauchoride/controllers/RideController.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    public Ride getByIdForUser(
+            @ApiParam("id") @RequestParam Long id) {
+        Ride ride = rideRepository.findByIdAndRiderId(id, getCurrentUser().getUser().getId())
+                .orElseThrow(() -> new EntityNotFoundException(Ride.class, id));
+
+        return ride;
+    }
+
+    @ApiOperation(value = "Create a new ride")
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/gauchoride/controllers/RideController.java:73]:
```suggestion
    @PreAuthorize("hasRole('ROLE_RIDER')")
```

**Review by pconrad — CHANGES_REQUESTED** [review]
A couple of small access things, but overall looks good.  Very close to being mergeable!

File: src/main/java/edu/ucsb/cs156/gauchoride/controllers/RideController.java (hunk 4 above)
Comment (strongani) [reply src/main/java/edu/ucsb/cs156/gauchoride/controllers/RideController.java:73]: Currently, we do not have a ROLE_RIDER, and the only roles are admin, driver, and user, where an admin is both a driver and a user, and a driver is not a user. So, I use ROLE_USER as a placeholder for ROLE_RIDER, and once a ROLE_RIDER is made (or cherrypicked from somewhere else), changing these permissions of ROLE_USER to ROLE_RIDER would be a quick fix. So, as of now, for functionality, ROLE_USER is required.

**Review by SophiaWeiler — APPROVED** [review]
lgtm

**Pull request closed without merging**